src/main/webapp/app/pages/usageAnalysisPage/usage-analysis-utils.ts

```diff
@@ -0,0 +1,227 @@
+import {
+  UserStats,
+  UsageSummary,
+  UserOverviewUsage,
+  UserUsage,
+} from 'app/shared/api/generated/API';
+import { ToggleValue } from './UsageAnalysisPage';
+import {
+  TABLE_MONTH_FORMAT,
+  TABLE_DAY_FORMAT,
+  TABLE_YEAR_FORMAT,
+  USAGE_YEAR_DETAIL_TIME_KEY,
+  USAGE_MONTH_DETAIL_TIME_KEY,
+  USAGE_DAY_DETAIL_TIME_KEY,
+} from 'app/config/constants';
+import moment from 'moment';
+
+export type TimeGroupedUsageRecords = {
+  [USAGE_YEAR_DETAIL_TIME_KEY]: UsageRecord[];
+  [USAGE_MONTH_DETAIL_TIME_KEY]: UsageRecord[];
+  [USAGE_DAY_DETAIL_TIME_KEY]: UsageRecord[];
+};
+
+export type UsageRecord = {
+  resource: string;
+  userEmail: string;
+  maxUsageProportion: number;
+  usage: number;
+  time: string;
+  userId?: string;
+};
+
+export type UserOverviewUsageWithUsageTypes = UserOverviewUsage & {
+  yearUsage: Record<string, UserStats>;
+  monthUsage: Record<string, UserStats>;
+  dayUsage: Record<string, UserStats>;
+};
+
+export type UsageSummaryWithUsageTypes = UsageSummary & {
+  year: Record<string, Record<string, number>>;
+  month: Record<string, Record<string, number>>;
+  day: Record<string, Record<string, number>>;
+};
+
+function checkIfUserUsage(
+  incomingData: UsageSummaryWithUsageTypes | UserOverviewUsageWithUsageTypes[]
+): incomingData is UserOverviewUsageWithUsageTypes[] {
+  return Array.isArray(incomingData);
+}
+
+function createResourceTableRecord(
+  time: string,
+  usageSummary: Record<string, number>,
+  keyType: 'resource' | 'email',
+  isAllUsage: boolean
+): UsageRecord[] {
+  return Object.entries(usageSummary)
+    .filter(([key]) => isAllUsage || !key.startsWith('/api/private/'))
```

Comment:
I noticed `key.startsWith('/api/private/'))` used a lot. I know it looks like a simple logic, but can we wrap it up as a method?